File: src/main/java/leetcode/topinterview/LeetCode101.java
package leetcode.topinterview;

import common.TreeNode;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/23 17:51
 */
public class LeetCode101 {


    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        return isSymmetric(root.left, root.right);
    }

    private boolean isSymmetric(TreeNode left, TreeNode right) {

        if (left == null && right == null) {
            return true;
        }
        if (left == null) {
            return false;
        }
        if (right == null) {
            return false;
        }
        if (left.val != right.val) {
            return false;
        }

        return isSymmetric(left.left, right.right) && isSymmetric(left.right, right.left);

    }
}

File: src/main/go/leetcode/lists/hot100/LeetCode101.go
package hot100

func isSymmetric(root *TreeNode) bool {
	var dfs func(l, r *TreeNode) bool
	dfs = func(l, r *TreeNode) bool {
		if l == nil && r == nil {
			return true
		}
		if l == nil && r != nil || r == nil && l != nil || l.Val != r.Val {
			return false
		}
		return dfs(l.Left, r.Right) && dfs(l.Right, r.Left)
	}

	return dfs(root.Left, root.Right)
}
